source/lac/dynamic_sparsity_pattern.cc

```diff
@@ -554,6 +554,47 @@ DynamicSparsityPattern::n_nonzero_elements() const
 }
 
 
+
+IndexSet
+DynamicSparsityPattern::nonempty_cols() const
+{
+  std::set<types::global_dof_index> cols;
+  for (const auto &line : lines)
+    cols.insert(line.entries.begin(), line.entries.end());
+
+  IndexSet res(this->n_cols());
+  res.add_indices(cols.begin(), cols.end());
+  return res;
+}
+
+
+
+IndexSet
+DynamicSparsityPattern::nonempty_rows() const
+{
+  const IndexSet  all_rows = complete_index_set(this->n_rows());
+  const IndexSet &rowset_  = rowset.size() == 0 ? all_rows : rowset;
+
+  std::vector<types::global_dof_index> rows;
+  auto                                 line = lines.begin();
+  AssertDimension(rowset_.n_elements(), lines.size());
+  for (const auto &row : rowset_)
+    {
+      if (line->entries.size() > 0)
+        rows.push_back(row);
+
+      ++line;
+    }
+
+  Assert(line == lines.end(), ExcInternalError());
```

Comment:
You've already checked that lines has the right size above -- no need to do it again here, and then you can also get rid of the `line` variable.

Reply:
will remove

> then you can also get rid of the line variable.

I don't think so.

Reply:
Right, I missed the other use of the variable.